mirrord/agent/src/file.rs

```diff
@@ -516,4 +531,65 @@ impl FileManager {
         })
         .map_err(ResponseError::from)
     }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn open_dir(&mut self, fd: u64) -> RemoteResult<OpenDirResponse> {
+        trace!("FileManager::open_dir -> fd {:#?}", fd,);
+        let path = match self
+            .open_files
+            .get(&fd)
+            .ok_or(ResponseError::NotFound(fd))?
+        {
+            RemoteFile::Directory(path) => path,
+            _ => panic!("Opened path is a file."),
+        };
+
+        let fd = self.index_allocator.next_index().ok_or_else(|| {
+            ResponseError::AllocationFailure("FileManager::open_relative".to_string())
+        })?;
+
+        let dir_stream = path.read_dir()?.enumerate();
+        self.dir_streams.insert(fd, dir_stream);
+
+        Ok(OpenDirResponse { fd })
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn read_dir(&mut self, fd: u64) -> RemoteResult<ReadDirResponse> {
+        let dir_stream = self
+            .dir_streams
+            .get_mut(&fd)
+            .ok_or(ResponseError::NotDirectory(fd))?;
+
+        let result = if let Some((offset, entry)) = dir_stream.next() {
+            let entry = entry?;
+
+            let mode = entry.metadata()?.mode();
+
+            let file_type = match mode & libc::S_IFMT {
+                libc::S_IFLNK => libc::DT_LNK,
+                libc::S_IFREG => libc::DT_REG,
+                libc::S_IFBLK => libc::DT_BLK,
+                libc::S_IFDIR => libc::DT_DIR,
+                libc::S_IFCHR => libc::DT_CHR,
+                libc::S_IFIFO => libc::DT_FIFO,
+                libc::S_IFSOCK => libc::DT_SOCK,
+                _ => libc::DT_UNKNOWN,
+            };
+
+            let internal_entry = DirEntryInternal {
+                inode: entry.ino(),
+                position: offset as u64,
+                name: entry.file_name().into_string().unwrap().into_bytes(),
```

Comment:
Pass it as String internally then convert it in layer to the proper encoding. (also unwrap is evil, return error.)